File: internal/middleware/access.go
package middleware

import (
	"context"

	"planner/internal/db"
)

// AccessError is returned by CanAccess when the user lacks permission.
type AccessError struct {
	Status  int
	Message string
}

func (e *AccessError) Error() string {
	return e.Message
}

// CanAccess checks whether userID has at least the required permission on
// plannerId. It returns the resolved level ("owner","edit","view") or an
// *AccessError with the appropriate HTTP status.
func CanAccess(ctx context.Context, database *db.DB, plannerID, userID int, require string) (string, error) {
	var ownerID int
	err := database.QueryRowContext(ctx,
		database.Rebind("SELECT owner_id FROM planners WHERE id = ?"), plannerID,
	).Scan(&ownerID)
	if err != nil {
		return "", &AccessError{Status: 404, Message: "Planner not found"}
	}

	if ownerID == userID {
		return "owner", nil
	}
	if require == "owner" {
		return "", &AccessError{Status: 403, Message: "Only the owner can do this"}
	}

	// Direct per-user share
	var perm string
	err = database.QueryRowContext(ctx,
		database.Rebind("SELECT permission FROM planner_shares WHERE planner_id = ? AND user_id = ?"),
		plannerID, userID,
	).Scan(&perm)
	if err == nil {
		if perm != "view" && perm != "edit" {
			return "", &AccessError{Status: 500, Message: "Invalid permission value in database"}
		}
		if require == "edit" && perm != "edit" {
			return "", &AccessError{Status: 403, Message: "Edit access required"}
		}
		return perm, nil
	}

	// Group-based access: find best permission across all groups the user belongs to.
	// COALESCE prefers per-member override over the group default.
	rows, qErr := database.QueryContext(ctx,
		database.Rebind(`SELECT COALESCE(pgmo.permission, pgs.default_permission) AS permission
		FROM planner_group_shares pgs
		JOIN group_members gm ON gm.group_id = pgs.group_id AND gm.user_id = ?
		LEFT JOIN planner_group_member_overrides pgmo
		  ON pgmo.planner_id = pgs.planner_id
		 AND pgmo.group_id   = pgs.group_id
		 AND pgmo.user_id    = ?
		WHERE pgs.planner_id = ?`),
		userID, userID, plannerID,
	)
	if qErr == nil {
		defer rows.Close()
		var best string
		for rows.Next() {
			var p string
			if scanErr := rows.Scan(&p); scanErr != nil {
				continue
			}
			if p == "edit" {
				best = "edit"
			} else if best == "" {
				best = "view"
			}
		}
		if best != "" {
			if require == "edit" && best != "edit" {
				return "", &AccessError{Status: 403, Message: "Edit access required"}
			}
			return best, nil
		}
	}

	return "", &AccessError{Status: 403, Message: "Access denied"}
}

File: server/middleware/access.ts
import { Response } from 'express';
import { query } from '../db';

/** Resolve the calling user's access level to a planner, or throw a {status, message} error. */
export async function canAccess(
  plannerId: number,
  userId: number,
  require: 'view' | 'edit' | 'owner'
): Promise<'owner' | 'edit' | 'view'> {
  const { rows } = await query<{ owner_id: number }>(
    'SELECT owner_id FROM planners WHERE id = $1', [plannerId]
  );
  if (!rows.length) throw { status: 404, message: 'Planner not found' };

  if (rows[0].owner_id === userId) return 'owner';
  if (require === 'owner') throw { status: 403, message: 'Only the owner can do this' };

  // Direct per-user share
  const { rows: shares } = await query<{ permission: string }>(
    'SELECT permission FROM planner_shares WHERE planner_id=$1 AND user_id=$2', [plannerId, userId]
  );
  if (shares.length) {
    const perm = shares[0].permission;
    if (perm !== 'view' && perm !== 'edit') throw { status: 500, message: 'Invalid permission value in database' };
    if (require === 'edit' && perm !== 'edit') throw { status: 403, message: 'Edit access required' };
    return perm;
  }

  // Group-based access: find best permission across all groups the user belongs to
  const { rows: groupPerms } = await query<{ permission: string }>(
    `SELECT COALESCE(pgmo.permission, pgs.default_permission) AS permission
     FROM planner_group_shares pgs
     JOIN group_members gm ON gm.group_id = pgs.group_id AND gm.user_id = $2
     LEFT JOIN planner_group_member_overrides pgmo
       ON pgmo.planner_id = pgs.planner_id
      AND pgmo.group_id   = pgs.group_id
      AND pgmo.user_id    = $2
     WHERE pgs.planner_id = $1`,
    [plannerId, userId]
  );

  if (groupPerms.length) {
    const perm = groupPerms.some(r => r.permission === 'edit') ? 'edit' : 'view';
    if (require === 'edit' && perm !== 'edit') throw { status: 403, message: 'Edit access required' };
    return perm;
  }

  throw { status: 403, message: 'Access denied' };
}

export function sendError(res: Response, status: number, message: string): void {
  res.status(status).json({ error: message });
}

export function handleRouteError(res: Response, err: unknown): void {
  const e = err as { status?: number; message?: string };
  if (e.status) { sendError(res, e.status, e.message || 'Error'); return; }
  console.error(err);
  sendError(res, 500, 'Internal server error');
}
